gh-132470: Prevent crash in ctypes.CField when `byte_size` is incorrect

dura0ok opened the pull request:
When creating a ctypes.CField with an incorrect byte_size (e.g., using byte_size=2 for ctypes.c_byte), the code would previously abort due to the failed assertion byte_size == info->size.

This commit replaces the assertion with a proper error handling mechanism that raises a ValueError when byte_size does not match the expected type size. This prevents the crash and provides a more informative error message to the us




<!-- gh-issue-number: gh-132470 -->
* Issue: gh-132470
<!-- /gh-issue-number -->

dura0ok committed:
Fix: Prevent crash in ctypes.CField when byte_size does not match type size  (gh-132470)

When creating a ctypes.CField with an incorrect byte_size (e.g., using byte_size=2 for ctypes.c_byte), the code would previously abort due to the failed assertion byte_size == info->size.

This commit replaces the assertion with a proper error handling mechanism that raises a ValueError when byte_size does not match the expected type size. This prevents the crash and provides a more informative error message to the us

sobolevn reviewed (commented):
Please, add a test case somewhere to `test_ctypes` and a NEWS entry.

brianschubert commented on Lib/test/leakers/test_ctypes.py:
```diff
@@ -13,3 +13,17 @@ class RECT(Structure):
     gc.collect()
+
+
+def test_invalid_byte_size_raises(self):
```
I'm not sure `Lib/test/leakers` is the right place for this test. This should probably go with the ctypes-proper tests in `Lib/test/test_ctypes`. `Lib/test/test_ctypes/test_struct_fields.py` might be a good fit.

dura0ok replied:
yep, moved, thanks.

picnixz commented on Misc/NEWS.d/next/C_API/2025-04-13-20-22-14.gh-issue-132470.uUUrUJ.rst:
```diff
@@ -0,0 +1,4 @@
+Creating a `ctypes.CField` with a `byte_size` that does not match the actual
```
No need for a NEWS entry. The structure is currently opaque and only exposed as a read-only structure. Users shouldn't be able to call the constructor directly.

dura0ok replied:
fixed

picnixz commented on Misc/NEWS.d/next/C_API/2025-04-13-20-23-59.gh-issue-132470.uUUrUJ.rst:
```diff
@@ -0,0 +1,4 @@
+Creating a `ctypes.CField` with a `byte_size` that does not match the actual
```
Remove this NEWs entry as well.

dura0ok replied:
fixed

dura0ok committed:
Add regression test for invalid byte_size in ctypes.CField

Adds a test to ensure that creating a ctypes.CField with a byte_size
that doesn't match the underlying C type size (e.g., 2 bytes for c_byte)
correctly raises a ValueError. This test verifies the fix for gh-132470
and prevents future regressions where such mismatches could cause an abort.

picnixz commented on Lib/test/test_ctypes/test_struct_fields.py:
```diff
@@ -75,6 +75,19 @@ def __init_subclass__(cls, **kwargs):
+                _internal_use=True
+            )
+
+            self.assertIn("does not match type size", str(cm.exception))
```
I suggest using `assertRaisesRegex` instead here.

dura0ok replied:
fixed

picnixz commented on Lib/test/test_ctypes/test_struct_fields.py:
```diff
@@ -75,6 +75,19 @@ def __init_subclass__(cls, **kwargs):
+            CField(
+                name="a",
+                type=c_byte,
+                byte_size=2, # Wrong size: c_byte is only 1 byte
```
```suggestion
                byte_size=2,  # Wrong size: c_byte is only 1 byte
```

dura0ok replied:
fixed

ZeroIntensity replied:
Hmm, did the commit get lost somewhere? It looks like this is still one space instead of two.

dura0ok committed:
add news

dura0ok committed:
fix review test

ZeroIntensity commented on Misc/NEWS.d/next/C_API/2025-04-13-20-52-39.gh-issue-132470.UqBQjN.rst:
What on earth happened to your news entry? Just redo this using `blurb_it`, please.

sobolevn replied:
@ZeroIntensity We can always suggest an impovment here on github :)

Many new contributors might not know how to work with cpython-specific tooling and what NEWS even are.

Here's my suggestion: https://github.com/python/cpython/pull/132475#discussion_r2046355923

ZeroIntensity replied:
Thanks for that, I didn't mean to come off as rude :). In hindsight, I didn't see the "first time contributor" mark, so that's my bad!

sobolevn commented on Misc/NEWS.d/next/C_API/2025-04-13-20-52-39.gh-issue-132470.UqBQjN.rst:
```diff
@@ -0,0 +1,14 @@
+# Write your Misc/NEWS.d entry below.  It should be a simple ReST paragraph.
+# Don't start with "- Issue #<n>: " or "- gh-issue-<n>: " or that sort of
+stuff.
+###########################################################################
```
```suggestion
Creating a :class:`ctypes.CField` with a *byte_size* that does not match the actual
type size now raises a :exc:`ValueError` instead of crashing the interpreter.
```

dura0ok committed:
Update news

Co-authored-by: sobolevn <mail@sobolevn.me>

sobolevn reviewed (approved):
Thank you!

ZeroIntensity reviewed (approved):
LGTM as well, but the formatting change that Bénédikt suggested seemingly got un-fixed.

That, and I didn't realize we had `skip news` on this. I'm going to remove that because I'd argue this is public enough for a news entry, but someone is free to bicker with me on that.

dura0ok committed:
fix space

sobolevn reviewed (approved):
Thanks! I will backport this to 3.13 as well.
